### ISSUE TITLE
AJAX action loading and inline section edit

### ISSUE THREAD

**Issue by zioth** [splitbrain/dokuwiki#1203]
Let's modernize Dokuwiki!

In the template I'm working on, I've overridden some of the core Doku JS to allow many actions to be fetched using AJAX instead of reloading the page. This makes Doku feel much more responsive. In particular, if you're viewing a page, you don't have to reload to switch to the edit, subscribe or revisions mode. Canceling an edit is instantaneous, and preview loads much faster as well.

In addition, I've made it so that section edits happen in place -- the section you're editing gets replaced with an editor, while the rest of the page is unchanged.

My code overrides existing JS, and adds an additional custom action (called "partial"), and depends on certain class names being attached to the body tag, so it can't be patched directly into Doku, but I'd be happy to share the code, and might even have time to make the change myself.

This feature would have to be based on a config option, since it might not play well with some plugins. In particular, it will mess up plugins which let you modify UI outside the main content area through ~~XXX~~ - style directives.

**Comment by Klap-in** [comment]
Could you share your fork at Github? It would be nice to have a look at your solution direction.

Do you have already examples of plugins which are affected?

**Comment by zioth** [comment]
It's more of a graft than a fork right now. I didn't want to mess with core Doku code, so I did it in my template. I tried to attach the files here, but only images are supported.

Maybe I'll make a plugin, which can later (possibly) be incorporated into Doku core.

**Comment by gamma** [comment]
It might not be the best way to do it, but I also already published a plugin to edit sections without leaving the page. Would be great to see your solution as this topic is something many are looking forward to.

https://github.com/i-net-software/dokuwiki-plugin-sectionedit

**Comment by zioth** [comment]
Thanks for the link. I have all the JS in place, but I didn't know how to intercept AJAX events.

**Comment by gamma** [comment]
That would be done best with an action component of a plugin.

**Comment by zioth** [comment]
I know. :) I was thanking you for giving me the code to do it, from your plugin.

**Comment by zioth** [comment]
I'm close to finishing the plugin. Is there a way in an action plugin to prevent the template from being called? I want a do= action where the template doesn't get involved at all. ACTION_ACT_PREPROCESS renders the template even if I call preventDefault().

**Comment by Chris--S** [comment]
If you are using lib/exe/ajax.php as an entry point, then you can use DOKU_AJAX_UNKNOWN and avoid act_dispatch() altogether.

**Comment by zioth** [comment]
True. I'll do that if I have to, but my plugin will be far simpler if I don't have to use ajax.php, since I can use the existing url handlers instead of adding a new id parameter and setting all the doku variables, to prepare for a call to tpl_content().

**Comment by Chris--S** [comment]
This is the url handlers in doku.php?  Would it be better to bundle the url handlers somewhere where they can be called by doku.php and then become available for use from ajax.php and plugins?

I say that as right now, there seems to be more than just calling the template in act_dispatch which would seem to get in the way of using it for an ajax response.

**Comment by zioth** [comment]
Yeah. It looked like that to me too. Ah well. I'll see how hard it is to do this in ajax. It was a lot easier to do it in my template, where I could intercept the action before rendering anything.

**Comment by Chris--S** [comment]
If you do put up a pull request with the refactoring of doku.php to make the url handlers callable, please @ reference me in the PR.

**Comment by zioth** [comment]
Looks like this can't easily be done as an AJAX plugin, because I'd have to copy most of doku.php. I tried including doku.php and canceling DOKUWIKI_STARTED, but that doesn't work. I could try canceling every render event, but there isn't any event to prevent the template from rendering.

It might have to be a branch of dokuwiki, which I'm not sure I want to get into.

**Comment by Chris--S** [comment]
Could you not move the bits of doku.php you're after out into a file called globals.php and include it in doku.php and also lib/exe:ajax.php

A simple refactor PR with good justification should be easy to get agreed and merged.

**Comment by zioth** [comment]
Good idea. I was just hoping my plugin could work with existing versions of doku.

**Comment by zioth** [comment]
I figured it out. I created a blank template, and pointed $conf['template'] at it.

**Comment by Chris--S** [comment]
I had thought of something similar, I don't think its a very good solution - except as a compatibility fix for older versions of dokuwiki.  I think it would be better to adapt ajax.php so it can properly be used for this sort of functionality.

**Comment by zioth** [comment]
My first goal is to get the plugin out there so people can use it. Afterwards, I'll see if I have time to refactor doku.php.

**Comment by gamma** [comment]
I ran into similar problems with https://github.com/i-net-software/dokuwiki-plugin-sectionedit - In the end I decided to register for ACTION_ACT_PREPROCESS, include doku.php and then have the action call tpl_content() and exit the process. Worked fine for me. The only quirky thing is: you have to do two AJAX calls, one for saving the changes and one for Fertigung the new content.

**Comment by zioth** [comment]
I've finished my plugin. It works with most of the most popular templates.

https://www.dokuwiki.org/plugin:fastwiki

**Comment by zioth** [comment]
Integrating this into doku core would be difficult, since it depends either on templates being set up in a particular way, or on custom per-template initialization code. It could be integrated in such a way that it detects supported templates.

And any template devs reading this -- it's easy to support my plugin. Take a look!

**Comment by gamma** [comment]
Is there a demo somewhere?

**Comment by zioth** [comment]
http://zioth.com/wiki/playground2

Username: demo. Password: demo.

You can log in using the door icon near the upper right.

**Comment by gamma** [comment]
Cool. Could you come up with a solution to save using AJAX as well?

**Comment by zioth** [comment]
That's much harder, since there are plugins (especially ~~WHATEVER~~ style plugins) that modify the page in ways I can't account for, and editing can modify the TOC, which I don't currently update

Maybe I'll give it a shot at some point.

**Comment by zioth** [comment]
That crossed-out WHATEVER was supposed to be surrounded by two tildes. Those kinds of plugins, and action plugins, might modify the page outside the main content area.

**Comment by zioth** [comment]
Your wish is (partially) my command! Saving now uses AJAX. Saving sections currently reloads the entire content area instead of just the section you edited, but that's on my todo list.

**Comment by gamma** [comment]
Awesome. I think replacing the whole content is a good thing, because you need to update the editing position for each edit button as well I guess. What do you do about JavaScript things that need to be registered on the new content, or do you gracefully skip this?

**Comment by zioth** [comment]
Thanks for pointing that out. I was mostly done, but I'd need some way to generate all range inputs, which would be a pain. Maybe I'll do it anyway.

What kinds of JavaScript things do you mean? Any embedded JS should be executed (though I should test it). As for other plugins and templates which hook into parts of the page content, there isn't really anything I can do about that except provide an API that developers can use. I already have one for templates. I should add one for plugins.

**Comment by gamma** [comment]
Yep, an API would be best. I think it would be great for DW to provide a way to DWishly register things to be done on startup, keep a reference and re-do them when editing is done. jQuery currently Spions this by removing all registered startup handlers.

I made a few plugins that rely on JS initialization, so my version of the sectionedit cheats by removing the current js.php from the DOM and injecting a new script with the js.php. It gets reloaded lazily and boom all the DW-JS things are registered again. Yes, this is evil cheating, but serves the purpose on inline editing.

**Comment by zioth** [comment]
If you do reload js.php, you'll end up with events doubly-attached to things. It could cause strange behavior.

**Comment by gamma** [comment]
Yes it could strangely it did not. Dunno if I de registered. But all events just canes as they should and it does not matter because you will reload the page eventually. Anyway: an initialization handler would be preferable.

**Comment by zioth** [comment]
I've added AJAXed page id changes, and pre-loading of links on the page, for instantaneous navigation on some wikis.

Given the variety of templates and plugins, this feature can only be integrated into core Dokuwiki if it's turned off by default, and limitations are imposed on which future templates and plugins are considered to support the latest version of DokuWiki. If this goes beyond what's practical for Doku, I'd be fine with closing this ticket and continuing discussion on https://www.dokuwiki.org/plugin:fastwiki
